bench-legacy/BinSearch.hs

```diff
@@ -0,0 +1,149 @@
+        --       have already explored.  This is a balancing act because of randomness in
+        --       execution time.
+        if good_trial time
+          then do
+            when verbose $
+              putStrLn
+                "[binsearch] Time in range.  LOCKING input size and performing remaining trials."
+            print_trial 1 n time
+            lockin (trials - 1) n [time]
+          else if time < 0.100
+                 then loop (2 * n)
+                 else do
+                   when verbose $
+                     putStrLn $
+                     "[binsearch] Estimated rate to be " ++
+                     show (round rate :: Integer) ++
+                     " per second.  Trying to scale up..."
+                        -- Here we've exited the doubling phase, but we're making our
+                        -- first guess as to how big a real execution should be:
+                   if time > 0.100 && time < 0.33 * desired_exec_length
+                     then do
+                       when verbose $
+                         putStrLn
+                           "[binsearch]   (Fudging first guess a little bit extra)"
+                       loop (round $ guess * initial_fudge_factor)
+                     else loop (round $ guess * fudge_factor)
+         -- Termination condition: Done with all trials.
+      lockin 0 n log = do
+        when verbose $
+          putStrLn $
+          "[binsearch] Time-per-unit for all trials: " ++
+          concat
+            (intersperse " " (map (show . (/ toDouble n) . toDouble) $ sort log))
+        return (n, log !! (length log `quot` 2)) -- Take the median
+      lockin trials_left n log = do
+        when verbose $
+          putStrLn
+            "[binsearch]------------------------------------------------------------"
+        time <- timeit $ kernel n
+                        -- hFlush stdout
+        print_trial (trials - trials_left + 1) n time
+                        -- whenverbose$ hFlush stdout
+        lockin (trials_left - 1) n (time : log)
+      print_trial :: Integer -> Integer -> NominalDiffTime -> IO ()
+      print_trial trialnum n time =
+        let rate = fromIntegral n / time
+            timeperunit = time / fromIntegral n
+         in when verbose $
+            putStrLn $
+            "[binsearch]  TRIAL: " ++
+            show trialnum ++
+            " secPerUnit: " ++
+            showTime timeperunit ++
+            " ratePerSec: " ++ show rate ++ " seconds: " ++ showTime time
+  (n, t) <- loop 1
+  return (n, fromRational $ toRational t)
+
+
+showTime ::  NominalDiffTime -> String
+showTime t = show ((fromRational $ toRational t) :: Double)
+
+toDouble :: Real a => a -> Double
+toDouble = fromRational . toRational
+
+
+-- Could use cycle counters here.... but the point of this is to time
+-- things on the order of a second.
+timeit :: IO () -> IO NominalDiffTime
+timeit io = do
+  strt <- getCurrentTime
```

Comment:
you might consider using [getMonotonicTimeNSec](https://hackage.haskell.org/package/base-4.14.0.0/docs/GHC-Clock.html#v:getMonotonicTimeNSec)

Reply:
I think that besides reformatting we made very little changes to legacy benchmarks. This is a good suggestion that requires little effort, it'll get updated